app/src/main/java/org/piwigo/data/repository/ImageRepository.java

```diff
@@ -327,30 +215,130 @@ public void updateImages(Integer categoryId) {
                         VariantWithImage vwi = new VariantWithImage();
                         vwi.image = i;
                         vwi.variant = new ImageVariant(i.id, i.width, i.height, location, null, d.url);
-                        Log.d("m_cache_sync","New variant "+vwi.image.name);
-                        return i;
+                        return new PositionedItem<VariantWithImage>(counter, vwi, true);
                     } else {
-                        Log.d("m_cache_sync","Existing variant "+existingVariant.image.name);
-                        return i;
+                        return new PositionedItem<>(counter, existingVariant, false);
                     }
                 });
-
-        restImages.subscribe();
-
-        Flowable<Image> dbImage = db.imageDao().getImagesInCategory(categoryId)
+        if(db == null) {
+            return remotes.toObservable();
+        }
+        return db.variantDao().getVariantsWithImageInCategory(categoryId)
             .subscribeOn(ioScheduler)
             .observeOn(ioScheduler)
             .flattenAsFlowable(s -> s)
+            .filter(variantWithImage -> {
+                if(!remoteIDs.contains(variantWithImage.image.id).blockingGet()) {
+                    Log.d("m_cache_sync","Deleted image "+variantWithImage.image.name);
+                    db.imageCategoryMapDao().deleteFromImage(variantWithImage.image.id);
+                    db.imageDao().delete(variantWithImage.image);
+                    return false;
+                }
+                return true;
+            })
             .zipWith(Flowable.range(0, Integer.MAX_VALUE),
-                (item, counter) -> {
-                    if(!restImages.contains(item).blockingGet()) {
-                        Log.d("m_cache_sync","Deleted image "+item.name);
-                        db.imageCategoryMapDao().deleteFromImage(item.id);
-                        db.imageDao().delete(item);
-                    }
-                    return item;
-                });
-        dbImage.subscribe();
+                (item, counter) -> new PositionedItem<VariantWithImage>(counter, item, true))
+            .toObservable();
```

Comment:
Don't we need a .concatWith(remotes) before .toObservable() here, as it is in the Categories?